File: solutions/median-sort/python/my_solution.py
# We import the sys library to make our program exits when it needs
import sys

# Check if the median given by the judge was not a '-1',
# which it means that the program has a problem and need to exits
def check_error(result:int) -> int:
    if result == -1:
        sys.exit()
    return result

# Prints the final result to the judge and verifies the code it gives
def check_answer(result) -> None:
    print(" ".join(map(str, result)), flush=True)
    answer = int(input())
    check_error(answer)

# Display and flush the given question-query
def query(i:int, j:int, k:int) -> int:
    print(i, j, k, flush=True)
    # Waits for the judge answer
    result = int(input())   # Get the median of the query
    return check_error(result)

# Sort the list
def median_sort(i_list: list):
    result = [1, 2] # We start our result list by a predifined value of 1, 2 
                    # where we know that 1 < 2, a < b
    i_size = len(i_list)

    # Checks which position in our result array the 'number' should be
    for number in range(3, i_size + 1):
        left = 0    # From the beginning of the result array
        right = len(result) - 1 # From the last element in the result array 

        # If there is two elements, then it's sorted
        while right-left >= 1:
            # Assign indexes according to ternary search
            a = left + (right - left) // 3
            b = right - (right - left) // 3
            
            # Asks for the median
            median = query(result[a], result[b], number)

            # In the case that the median is a, (c,a,b)
            if median == result[a]:
                right = a - 1
                if left == right:  
                    right += 1

            # In the case that the median is b, (a,b,c)
            elif median == result[b]:
                left = b + 1
                if left == right:
                    left -= 1

            # In the case that the median is c, (a,c,b)
            elif median == number:
                left = a + 1
                right = b - 1
                if left == right:
                    left -= 1

        # Insert the given number at the leftist position
        result.insert(left, number)

    # Prints the result and checks if the judge gave an error code
    check_answer(result)

# Gets the parameters of the judge initial answer
T, N, Q = [int(initial_answer) for initial_answer in input().split()]

# Runs through all cases
for test_case in range(T):
    initial_list = range(1, N + 1)
    median_sort(initial_list)

# Exits when it finishes
sys.exit()

File: solutions/median-sort/haskell/my_solution.hs
import Control.Monad (foldM, replicateM)
import System.IO (hFlush, stdout)
import Data.Tuple (swap)
import Data.List.Split ( splitOn )
import System.Exit ( die )

-- Converts list with strings to list with numbers
parseToNumbers :: [String] -> [Int]
parseToNumbers = map read

checkError :: Int -> IO Int
checkError result = if result == -1 then System.Exit.die "-1" else return result

-- Print values and get the median of them
getMedian :: Show a => (a, a, a) -> IO Int
getMedian (a,b,c) = do
  let 
    raw_query = map show [a,b,c]
    query = unwords raw_query
  putStrLn query
  hFlush stdout
  answer <- readLn
  checkError answer

-- Sort the left part of the (a,b,c)
sortLeft :: Int -> Int -> Int -> IO (Int, Int)
sortLeft left a b = 
  median <$> getMedian (left, a, b) where
  median c | c == a = (a,b) 
           | c == b = (b,a)

-- Sort the right part of the (a,b,c)
sortRight :: Int -> Int -> Int -> IO (Int, Int)
sortRight right a b =
  median <$> getMedian (right, a, b) where
  median c | c == a = (b,a)
           | c == b = (a,b)

-- Order the values according to their position of the (left, center, right) recursively
order :: [Int] -> (Int, Int) -> IO [Int]
order [] (a, b) = return [a, b] -- Returns until there is no more values to check
order result (a, b) = 
  do
    (left, center, right) <- getPositionMedian (a, b) result -- Gets the position, we define a < b
    
    -- Orders until leftest value
    most_left <- case left 
      of
        (new_a:new_b:most_most_left) -> sortRight a new_a new_b >>= order most_most_left
        most_most_left -> return most_most_left
    
    -- Orders until centerest value
    most_center <- case center of
            (new_a:new_b:most_most_center) -> sortLeft a new_a new_b >>= order most_most_center
            most_most_center -> return most_most_center
    
    -- Orders until rightest value
    most_right <- case right of
            (new_a:new_b:most_most_right) -> sortLeft b new_a new_b >>= order most_most_right
            most_most_right -> return most_most_right
    
    -- Until it finishes returns the ordered result
    let result = most_left ++ a : most_center ++ b : most_right
    
    return result

-- Gets the position of the median (a,b,c) where c is the number we checked
getPositionMedian :: Foldable t => (Int, Int) -> t Int -> IO ([Int], [Int], [Int])
getPositionMedian (a, b) = foldM compareMedianResult ([], [], []) 
  where
    -- Gets what is the median given by the judge
    compareMedianResult (new_a, new_b, new_c) number = 
      median <$> getMedian (a, b, number) 
      where
        median m | m == a = (number:new_a, new_b, new_c)
                 | m == b = (new_a, new_b, number:new_c)
                 | m == number = (new_a, number:new_b, new_c)

-- Loops until all test cases are done
loopTestCase :: Int -> IO ()
loopTestCase index =
  -- We start by 3 because if there is only 2 numbers, the list is already sorted
  order [3..index] (1,2) >>=
  putStrLn  . unwords . map show >>
  hFlush stdout >>
  getLine >>
  return ()

main :: IO [()]
main = do
  -- Gets the judge initial input
  line1 <- getLine
  let 
    -- Saves the values in a int list
    values = splitOn " " line1
    int_values = parseToNumbers values
    -- Assign the values in a variable
    t = head int_values
    n = int_values !! 1
    q = int_values !! 2
  -- Starts the sorted according to the test cases given
  replicateM t (loopTestCase n)